File: backend-crud/Backend/Startup.fs
namespace Backend

open System
open Microsoft.AspNetCore.Builder
open Microsoft.AspNetCore.Hosting
open Microsoft.AspNetCore.Http
open Microsoft.Extensions.DependencyInjection
open Microsoft.Extensions.Hosting

open RebelSoftware.LoggingService
open RebelSoftware.MessageQueueService
open RebelSoftware.SerializationService
open RebelSoftware.HttpService

type Startup() =

    // This method gets called by the runtime. Use this method to add services to the container.
    // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
    member this.ConfigureServices(services: IServiceCollection) =
        ()

    // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
    member this.Configure(app: IApplicationBuilder, env: IWebHostEnvironment) =
        let logger = Logging.createLogger ()
        let messageQueuer = MessageQueueing.createMessageQueuer ()
        let serializationService = Serialization.createSerializationService ()
        let getHttpService httpContext = 
            Http.createHttpServer httpContext
        let employeeValidator = Validation.getEmployeeValidator ()        

        if env.IsDevelopment() then
            app.UseDeveloperExceptionPage() |> ignore

        app.UseRouting() |> ignore
        
        app.UseEndpoints(fun endpoints ->
            endpoints.MapPost("/employee/create", fun context -> EmployeeController.create logger messageQueuer serializationService (getHttpService context) employeeValidator) |> ignore
            //endpoints.MapPost("/employee/create", fun context -> Employee.create_dummy context)
              //todo: why do we get a compile error here if this is not a lambda? Can't see why the 2nd param can't just be RandomNumbers.insertOne, instead of a lambda that calls RandomNumbers.insertOne
            //todo: use generic exception page, for when we catch an Exception while writing a response                 
            ) |> ignore  

File: backend-crud-CSharp/backend-crud-CSharp/Startup.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Hosting;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.HttpsPolicy;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Hosting;

using RebelSoftware.LoggingService;
using RebelSoftware.MessageQueueService;
using RebelSoftware.SerializationService;
using RebelSoftware.HttpService;

namespace backend_crud_CSharp
{
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddRazorPages();
            services.AddCors();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            //var logger = RebelSoftware.LoggingService.Logging.createLogger();
            var logger = Logging.createLogger();
            var messageQueuer = MessageQueueing.createMessageQueuer();
            var serializationService = Serialization.createSerializationService<Model.Employee>();
            var employeeValidator = Validation.getEmployeeValidator();


            if (env.IsDevelopment())
            {
                //app.UseDeveloperExceptionPage();
            }
            else
            {
                //app.UseExceptionHandler("/Error");
                // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                //app.UseHsts();
            }

            //app.UseHttpsRedirection();
            //app.UseStaticFiles();

            app.UseRouting();

            app.UseCors(builder => builder.AllowAnyMethod()
                              .AllowAnyHeader()
                              .SetIsOriginAllowed(origin => true)
                              .AllowCredentials());

            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapRazorPages();
                endpoints.MapPost("/employee/create", async context => 
                {
                    var httpService = HttpServiceFactory.CreateHttpService(context);
                    var controller = new EmployeeController(logger, messageQueuer, serializationService, httpService, employeeValidator);

                    controller.Create();
                    //await context.Response.WriteAsync("hello");
                });
            });
        }
    }
}
